File: 1428-jump-game-iii/solution.java
class Solution {
    int[] arr;
    int[] visited;
    Map<Integer, Boolean> map = new HashMap<>();

    public boolean canReach(int[] arr, int start) {
        this.arr = arr;
        this.visited = new int[arr.length];

        return recursion(start);
    }   

    private boolean recursion(int index){
        boolean exp = (0 <= index && index < arr.length);
        
        if(!exp || visited[index] == 1){
            return false;
        }

        if(map.containsKey(index)){
            return map.get(index);
        }

        if(arr[index] == 0){
            return true;
        }
        visited[index] = 1;
        boolean res = recursion(index + arr[index]) || recursion(index - arr[index]);
        visited[index] = 0;
        map.put(index, res);
        return res;
    }
}

File: 1428-jump-game-iii/solution.rs
impl Solution {
    pub fn can_reach(arr: Vec<i32>, start: i32) -> bool {
        let mut visited: Vec<bool> = Vec::new();
        
        for i in 0..arr.len() {
            visited.push(false);
        }

        return Solution::recurse(&arr, start, &mut visited);
    }

    pub fn recurse(arr: &Vec<i32>, start: i32, visited: &mut Vec<bool>) -> bool {
        if start > (arr.len() - 1) as i32 || 0 > start {
            return false;
        }

        if visited[start as usize] {
            return false;
        }

        if arr[start as usize] == 0 {
            return true;
        }

        visited[start as usize] = true;

        return Solution::recurse(
            arr,
            start + arr[start as usize],
            visited
        ) || Solution::recurse(
            arr, 
            start - arr[start as usize],
            visited
        );
    }
}
